service/history/replication/ack_manager.go

```diff
@@ -254,49 +258,65 @@ func (p *ackMgrImpl) GetTasks(
 
 func (p *ackMgrImpl) getTasks(
 	ctx context.Context,
+	pollingCluster string,
 	minTaskID int64,
 	maxTaskID int64,
-	batchSize int,
 ) ([]*replicationspb.ReplicationTask, int64, error) {
 	if minTaskID > maxTaskID {
-		return nil, 0, serviceerror.NewUnavailable("min task ID < max task ID, probably due to shard re-balancing")
+		return nil, 0, serviceerror.NewUnavailable("min task ID > max task ID, probably due to shard re-balancing")
 	} else if minTaskID == maxTaskID {
-		return []*replicationspb.ReplicationTask{}, maxTaskID, nil
+		return nil, maxTaskID, nil
 	}
 
-	replicationTasks := make([]*replicationspb.ReplicationTask, 0, batchSize)
-	iter := collection.NewPagingIterator(p.getPaginationFn(ctx, minTaskID, maxTaskID, batchSize))
-	for iter.HasNext() && len(replicationTasks) < batchSize {
+	replicationTasks := make([]*replicationspb.ReplicationTask, 0, p.pageSize())
+	skippedTaskCount := 0
+	lastTaskID := maxTaskID // If no tasks are returned, then it means there are no tasks bellow maxTaskID.
+	iter := collection.NewPagingIterator(p.getReplicationTasksFn(ctx, minTaskID, maxTaskID, p.pageSize()))
+	// iter.HasNext() should be the last check to avoid extra page read in case if replicationTasks is already full.
+	for len(replicationTasks) < p.pageSize() && skippedTaskCount <= p.maxSkipTaskCount() && iter.HasNext() {
 		task, err := iter.Next()
 		if err != nil {
-			p.logger.Error("replication task reader encounter error, return earlier", tag.Error(err))
-			if len(replicationTasks) == 0 {
-				return nil, 0, err
-			} else {
-				return replicationTasks, replicationTasks[len(replicationTasks)-1].GetSourceTaskId(), nil
-			}
+			return p.swallowPartialResultsError(replicationTasks, lastTaskID, err)
 		}
 
-		if replicationTask, err := p.toReplicationTask(ctx, task); err != nil {
-			p.logger.Error("replication task reader encounter error, return earlier", tag.Error(err))
-			if len(replicationTasks) == 0 {
-				return nil, 0, err
-			} else {
-				return replicationTasks, replicationTasks[len(replicationTasks)-1].GetSourceTaskId(), nil
+		// If, for any reason, task is skipped:
+		//  - lastTaskID needs to be updated because this task should not be read next time,
+		//  - skippedTaskCount needs to be incremented to prevent timeout on caller side (too many tasks are skipped).
+		// If error has occurred though, lastTaskID shouldn't be updated, and next time task needs to be read again.
+
+		ns, err := p.namespaceRegistry.GetNamespaceByID(namespace.ID(task.GetNamespaceID()))
+		if err != nil {
+			if _, isNotFound := err.(*serviceerror.NamespaceNotFound); !isNotFound {
```

Comment:
Is it possible that this could happen after host restart and before ns refresh fetched the namespace info?

Reply:
First call to load namespaces is from [registry.Start](https://github.com/temporalio/temporal/blob/997434cc5b35d88fa20e3b51766b0eb8dd9d57e9/common/namespace/registry.go#L229) method, which is blocking and called by `fx` during container is getting built.